compiler/src/Language/Bond/Codegen/Java/Class_java.hs

```diff
@@ -10,25 +10,362 @@ module Language.Bond.Codegen.Java.Class_java
+makeStructBuilderMember_makeGenericType declName declParams = [lt|
+            private #{typeParamAnglesNameList declParams} #{typeDescriptorNameWithParams declName declParams} makeGenericType(#{typeParamVarDeclList declParams}) {
+                #{newlineSepEnd 4 checkArg declParams}return #{castExpr} this.getInitializedFromCache(#{typeParamVarNameList declParams});
+            }
+|]
+    where
+        checkArg t@TypeParam {..} = [lt|#{ensureNotNullArgument (typeParamVarName t)};|]
+        castExpr = [lt|(StructBondTypeImpl)|]
+
+
+-- given struct class name and generic type parameters, builds text for implementation of
+-- the StructBondTypeBuilderImpl class which is responsible for building the type descriptor
+makeStructBondTypeMember_StructBondTypeBuilderImpl :: String -> [TypeParam] -> Text
+makeStructBondTypeMember_StructBondTypeBuilderImpl declName declParams = [lt|
+        static final class StructBondTypeBuilderImpl extends com.microsoft.bond.StructBondType.StructBondTypeBuilder<#{declName}> {
+            #{ifThenElse (null declParams) mempty (makeStructBuilderMember_makeGenericType declName declParams)}
+            @Override
+            public final int getGenericTypeParameterCount() {
+                return #{length declParams};
+            }
+
+            @Override
+            protected final #{typeDescriptorName declName} buildNewInstance(com.microsoft.bond.BondType[] genericTypeArguments) {
+                return new StructBondTypeImpl(#{ifThenElse (null declParams) "null" "new com.microsoft.bond.GenericTypeSpecialization(genericTypeArguments)"});
+            }
+
+            static void register() {
+                registerStructType(#{declName}.class, new StructBondTypeBuilderImpl());
+            }
+        }|]
+
+
+-- given generic type parameters, struct fields, and base type, builds text for implementation of
+-- the StructBondTypeImpl.initialize method
+makeStructBondTypeMember_initialize :: MappingContext -> [TypeParam] -> [Field] -> Maybe Type -> Text
+makeStructBondTypeMember_initialize java declParams structFields structBase = [lt|
+        @Override
+        protected final void initialize() {#{typeArgVarDeclList}#{fieldDescriptorInitList}
+            super.initializeBaseAndFields(#{baseTypeDescriptorParam}#{fieldTypeDescriptorParamsSeparator}#{fieldTypeDescriptorParams});
+        }|]
+    where
+        typeArgVarDeclList = newlineBeginSep 3 typeArgVarDecl indexedDeclParams
+            where
+                typeArgVarDecl (index, typeParam) = [lt|#{typeParamVarDecl typeParam} = this.getGenericSpecialization().getGenericTypeArgument(#{index});|]
+                indexedDeclParams = zip [0 :: Int ..] declParams
+        fieldDescriptorInitList = newlineBeginSep 3 fieldDescriptorInit structFields
+            where
+                fieldDescriptorInit Field {..} = [lt|this.#{fieldName} = new #{structFieldDescriptorTypeName java fieldType}(#{constructorParams});|]
+                    where
```

Comment:
The `where` clauses don't need to be nested each time. It's pretty common to write a `where` clause entry that refers to another where clause entry right below it. Here's a toy(ish) example:

    where
      fieldInits = basicInits ++ complexInits
      basicInits = map defaultInit basicFields
      complexInits = map explicitInit complexFields
      (complexFields,basicFields) = Data.List.partition hasExplicitDefault structFields